# Data Preprocessing

## Import dataset and analyze it

In [1]:
import pandas as pd
import zipfile

zip_path = "BindingDB_PDSPKi_202504_tsv.zip"
tsv_filename = "BindingDB_PDSPKi.tsv"

with zipfile.ZipFile(zip_path, 'r') as zip_ref:
    with zip_ref.open(tsv_filename) as file:
        original_df = pd.read_csv(file, sep='\t', on_bad_lines='skip')

original_df.head()

BindingDB Reactant_set_id   
0                     157753  \
1                     157754   
2                     157755   
3                     157756   
4                     157757   

                                       Ligand SMILES   
0  Cn1c2ncn(CCN3CCC(CC3)C(=O)c3ccc(F)cc3)c2c(=O)n...  \
1                           NC(N)=NN=Cc1c(Cl)cccc1Cl   
2  Fc1ccc(cc1)C(=O)C1CCN(CCn2c(=O)[nH]c3ccccc3c2=...   
3  Cc1nc2ccccn2c(=O)c1CCN1CCC(CC1)=C(c1ccc(F)cc1)...   
4  Cc1nc2sccn2c(=O)c1CCN1CCC(CC1)=C(c1ccc(F)cc1)c...   

                                        Ligand InChI   
0  InChI=1S/C21H24FN5O3/c1-24-19-17(20(29)25(2)21...  \
1  InChI=1S/C8H8Cl2N4/c9-6-2-1-3-7(10)5(6)4-13-14...   
2  InChI=1S/C22H22FN3O3/c23-17-7-5-15(6-8-17)20(2...   
3  InChI=1S/C29H27F2N3O/c1-20-26(29(35)34-16-3-2-...   
4  InChI=1S/C27H25F2N3OS/c1-18-24(26(33)32-16-17-...   

              Ligand InChI Key  BindingDB MonomerID   
0  PMEYQPKJAPXGPS-UHFFFAOYSA-N                81981  \
1  WDZVGELJXXEGPV-UHFFFAOYSA-N                81979   
2  FPCCSQOGAWCVBH-UHFFFAOYSA-N                21395   
3  ZGUPMFYFHHSNFK-UHFFFAOYSA-N                81978   
4  JUQLTPCYUFPYKE-UHFFFAOYSA-N             50001775   

                               BindingDB Ligand Name   
0  7-(2-(4-(4-fluorobenzoyl)piperidino)ethyl)theo...  \
1  CAS_5051-62-7::CAS_65636::Fluprofylline::Guana...   
2  3-(2-(4-(4-Fluorobenzoyl)piperidinol)ethyl)-2,...   
3                    CAS_159037::NSC_159037::R 56413   
4  (ritanserin)6-(2-{4-[Bis-(4-fluoro-phenyl)-met...   

                       Target Name   
0  5-hydroxytryptamine receptor 2A  \
1  5-hydroxytryptamine receptor 2A   
2  5-hydroxytryptamine receptor 2A   
3  5-hydroxytryptamine receptor 2A   
4  5-hydroxytryptamine receptor 2A   

  Target Source Organism According to Curator or DataSource Ki (nM)   
0                                  Rattus norvegicus            4.5  \
1                                  Rattus norvegicus          199.5   
2                                  Rattus norvegicus            3.2   
3                                  Rattus norvegicus              6   
4                                  Rattus norvegicus            5.5   

   IC50 (nM)  ...  UniProt (SwissProt) Recommended Name of Target Chain   
0        NaN  ...                    5-hydroxytryptamine receptor 2A     \
1        NaN  ...                    5-hydroxytryptamine receptor 2A      
2        NaN  ...                    5-hydroxytryptamine receptor 2A      
3        NaN  ...                    5-hydroxytryptamine receptor 2A      
4        NaN  ...                    5-hydroxytryptamine receptor 2A      

   UniProt (SwissProt) Entry Name of Target Chain   
0                                       5HT2A_RAT  \
1                                       5HT2A_RAT   
2                                       5HT2A_RAT   
3                                       5HT2A_RAT   
4                                       5HT2A_RAT   

   UniProt (SwissProt) Primary ID of Target Chain   
0                                          P14842  \
1                                          P14842   
2                                          P14842   
3                                          P14842   
4                                          P14842   

   UniProt (SwissProt) Secondary ID(s) of Target Chain   
0                                                NaN    \
1                                                NaN     
2                                                NaN     
3                                                NaN     
4                                                NaN     

   UniProt (SwissProt) Alternative ID(s) of Target Chain   
0                                                NaN      \
1                                                NaN       
2                                                NaN       
3                                                NaN       
4                                                NaN       

   UniProt (T

In [2]:
original_df.shape

(27694, 50)

In [3]:
original_df.columns

Index(['BindingDB Reactant_set_id', 'Ligand SMILES', 'Ligand InChI',
       'Ligand InChI Key', 'BindingDB MonomerID', 'BindingDB Ligand Name',
       'Target Name',
       'Target Source Organism According to Curator or DataSource', 'Ki (nM)',
       'IC50 (nM)', 'Kd (nM)', 'EC50 (nM)', 'kon (M-1-s-1)', 'koff (s-1)',
       'pH', 'Temp (C)', 'Curation/DataSource', 'Article DOI',
       'BindingDB Entry DOI', 'PMID', 'PubChem AID', 'Patent Number',
       'Authors', 'Institution', 'Link to Ligand in BindingDB',
       'Link to Target in BindingDB',
       'Link to Ligand-Target Pair in BindingDB', 'Ligand HET ID in PDB',
       'PDB ID(s) for Ligand-Target Complex', 'PubChem CID', 'PubChem SID',
       'ChEBI ID of Ligand', 'ChEMBL ID of Ligand', 'DrugBank ID of Ligand',
       'IUPHAR_GRAC ID of Ligand', 'KEGG ID of Ligand', 'ZINC ID of Ligand',
       'Number of Protein Chains in Target (>1 implies a multichain complex)',
       'BindingDB Target Chain Sequence', 'PDB ID(s) of Target

In [4]:
print(len(original_df["Target Name"].unique()))

398


In [5]:
# Find the most frequent value in "Target Name"
top_target = original_df["Target Name"].value_counts().idxmax()
top_count = original_df["Target Name"].value_counts().max()

print(f"Most frequent 'Target Name': {top_target} ({top_count} occurrences)")

Most frequent 'Target Name': D(2) dopamine receptor (1632 occurrences)


## Filter dataset
Only Ligand whose target is dopamine receptor will be considered. Input features are Ligand SMILES and output is Ki (nM)

In [6]:
# Filter the dataset
filtered_df = original_df[original_df["Target Name"] == top_target]
# Remove rows with missing values in the "Ki (nM)" column
filtered_df = filtered_df.dropna(subset=["Ki (nM)"])

# Check the shape and preview the filtered dataset
print(f"Filtered dataset has {filtered_df.shape[0]} rows.")
filtered_df.head()

Filtered dataset has 1632 rows.


BindingDB Reactant_set_id   
472                     154285  \
473                     154286   
474                     154287   
475                     154288   
476                     154289   

                                         Ligand SMILES   
472  CC(C)C[C@@H]1N2C(=O)[C@](NC(=O)[C@H]3CN(C)[C@@...  \
473                              NC1CCc2cc(O)c(O)cc2C1   
474  CC(C)C[C@@H]1N2C(=O)[C@](NC(=O)[C@H]3CN(C)C4Cc...   
475             CN1CCc2cccc-3c2[C@H]1Cc1ccc(O)c(O)c-31   
476             CN1CCc2cccc-3c2[C@H]1Cc1ccc(O)c(O)c-31   

                                          Ligand InChI   
472  InChI=1S/C32H40BrN5O5/c1-16(2)12-24-29(40)37-1...  \
473  InChI=1S/C10H13NO2/c11-8-2-1-6-4-9(12)10(13)5-...   
474  InChI=1S/C32H41N5O5/c1-17(2)12-25-29(39)36-11-...   
475  InChI=1S/C17H17NO2/c1-18-8-7-10-3-2-4-12-15(10...   
476  InChI=1S/C17H17NO2/c1-18-8-7-10-3-2-4-12-15(10...   

                Ligand InChI Key  BindingDB MonomerID   
472  OZVBMTJYIDMWIL-AYFBDAFISA-N                81993  \
473  ASXGAOFCKGHGMF-UHFFFAOYSA-N                81195   
474  YDOTUXAWKBPQJW-ISVRSENOSA-N                50855   
475  VMWNQDUVQKEIOC-CYBMUJFWSA-N             50001955   
476  VMWNQDUVQKEIOC-CYBMUJFWSA-N             50001955   

                                 BindingDB Ligand Name   
472  BROMOCRIPTINE::Bromocriptine+ (GTP+)::Bromocri...  \
473  (+/-)-2-Amino-6,7-dihydroxy-1,2,3,4-tetrahydro...   
474  ALPHA-ERGOCRYPTINE::MLS000069839::SMR000058703...   
475  (-)6-Methyl-5,6,6a,7-tetrahydro-4H-dibenzo[de,...   
476  (-)6-Methyl-5,6,6a,7-tetrahydro-4H-dibenzo[de,...   

                Target Name   
472  D(2) dopamine receptor  \
473  D(2) dopamine receptor   
474  D(2) dopamine receptor   
475  D(2) dopamine receptor   
476  D(2) dopamine receptor   

    Target Source Organism According to Curator or DataSource Ki (nM)   
472                                         Bos taurus            2.5  \
473                                         Bos taurus            120   
474                                         Bos taurus            1.3   
475                                         Bos taurus             51   
476                                         Bos taurus             43   

     IC50 (nM)  ...  UniProt (SwissProt) Recommended Name of Target Chain   
472        NaN  ...                             D(2) dopamine receptor     \
473        NaN  ...                             D(2) dopamine receptor      
474        NaN  ...                             D(2) dopamine receptor      
475        NaN  ...                             D(2) dopamine receptor      
476        NaN  ...                             D(2) dopamine receptor      

     UniProt (SwissProt) Entry Name of Target Chain   
472                                      DRD2_BOVIN  \
473                                      DRD2_BOVIN   
474                                      DRD2_BOVIN   
475                                      DRD2_BOVIN   
476                                      DRD2_BOVIN   

     UniProt (SwissProt) Primary ID of Target Chain   
472                                          P20288  \
473                                          P20288   
474                                          P20288   
475                                          P20288   
476                                          P20288   

     UniProt (SwissProt) Secondary ID(s) of Target Chain   
472                                                NaN    \
473                                                NaN     
474                                                NaN     
475                                                NaN     
476                                                NaN     

     UniProt (SwissProt) Alternative ID(s) of Target Chain   
472                                                NaN      \
473                                                NaN       
474                                                NaN       
475                                               

In [7]:
# Final dataset: keep only rows with non-null Ki values
final_df = filtered_df[["Ligand SMILES", "Ki (nM)"]].dropna()

# Check the result
final_df.head()

Ligand SMILES Ki (nM)
472  CC(C)C[C@@H]1N2C(=O)[C@](NC(=O)[C@H]3CN(C)[C@@...     2.5
473                              NC1CCc2cc(O)c(O)cc2C1     120
474  CC(C)C[C@@H]1N2C(=O)[C@](NC(=O)[C@H]3CN(C)C4Cc...     1.3
475             CN1CCc2cccc-3c2[C@H]1Cc1ccc(O)c(O)c-31      51
476             CN1CCc2cccc-3c2[C@H]1Cc1ccc(O)c(O)c-31      43

## Output features (Ki (nM)) processing

In [8]:
import numpy as np

# Convert to numeric, coerce errors to NaN
final_df["Ki (nM)_converted"] = pd.to_numeric(final_df["Ki (nM)"], errors='coerce')

# Filter out non-positive or NaN values
final_df_Ki_filtered = final_df[final_df["Ki (nM)_converted"].notna() & (final_df["Ki (nM)_converted"] > 0)]

# Add log-transformed Ki, since biochemical values span several orders of magnitude 
# and the log scale is often more informative:
final_df_Ki_filtered["log_Ki"] = np.log10(final_df_Ki_filtered["Ki (nM)_converted"])

df_wlogKi = final_df_Ki_filtered

# Check the result
print(df_wlogKi.shape)
df_wlogKi.head()

(1439, 4)


C:\Users\alice\AppData\Local\Temp\ipykernel_164024\3735405964.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df_Ki_filtered["log_Ki"] = np.log10(final_df_Ki_filtered["Ki (nM)_converted"])


Ligand SMILES Ki (nM)   
472  CC(C)C[C@@H]1N2C(=O)[C@](NC(=O)[C@H]3CN(C)[C@@...     2.5  \
473                              NC1CCc2cc(O)c(O)cc2C1     120   
474  CC(C)C[C@@H]1N2C(=O)[C@](NC(=O)[C@H]3CN(C)C4Cc...     1.3   
475             CN1CCc2cccc-3c2[C@H]1Cc1ccc(O)c(O)c-31      51   
476             CN1CCc2cccc-3c2[C@H]1Cc1ccc(O)c(O)c-31      43   

     Ki (nM)_converted    log_Ki  
472                2.5  0.397940  
473              120.0  2.079181  
474                1.3  0.113943  
475               51.0  1.707570  
476               43.0  1.633468

In [9]:
# Identify unselected (invalid) values
invalid_ki = final_df[(final_df["Ki (nM)_converted"].isna()) | (final_df["Ki (nM)_converted"] <= 0)]["Ki (nM)"]

# Visualize those values
print("Values that were not selected:")
print(invalid_ki.unique())

Values that were not selected:
['>10000']


Ki values that are '>10000' are not precise so they were not selected.

## Conversion to numpy + Input features (SMILES) processing

In [10]:
from rdkit import Chem
from rdkit.Chem import Descriptors
import numpy as np
from tqdm import tqdm

tqdm.pandas()

# Convert SMILES to Mol objects
mols = df_wlogKi["Ligand SMILES"].progress_apply(Chem.MolFromSmiles)

# Filter out invalid mols
valid_idx = mols.notnull()
mols = mols[valid_idx].reset_index(drop=True)
log_Ki = df_wlogKi.loc[valid_idx, "log_Ki"].reset_index(drop=True)

# Descriptor extraction to NumPy
def compute_descriptors_np(mol):
    return np.array([
        Descriptors.MolWt(mol),
        Descriptors.NumHDonors(mol),
        Descriptors.NumHAcceptors(mol),
        Descriptors.TPSA(mol),
        Descriptors.NumRotatableBonds(mol),
        Descriptors.RingCount(mol)
    ])

descriptor_list = [compute_descriptors_np(mol) for mol in tqdm(mols)]
X_desc = np.vstack(descriptor_list)         # shape: (n_samples, 6)
y = log_Ki.to_numpy().reshape(-1, 1)         # shape: (n_samples, 1)

100%|██████████| 1439/1439 [00:00<00:00, 2339.45it/s]


In [11]:
# Visualization
print(X_desc.shape)
print(X_desc)

(1439, 6)
[[654.606   3.      6.    118.21    5.      7.   ]
 [179.219   3.      3.     66.48    0.      2.   ]
 [575.71    3.      6.    118.21    5.      7.   ]
 ...
 [501.071   2.      5.     66.07    7.      4.   ]
 [501.071   2.      5.     66.07    7.      4.   ]
 [369.487   2.      6.    101.73    7.      2.   ]]


First, we will use X_desc (1439, 6) to analyze the baseline effectiveness of our models. Then we will add more feature engineering with Morgan fingerprints to increase the complexity of the data.

In [12]:
# Normalize X_desc
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_desc_scaled = scaler.fit_transform(X_desc)
print(X_desc_scaled)

[[ 3.6722227   1.74634788  1.46544034  2.1946403   0.41987974  3.12802546]
 [-2.17683836  1.74634788 -0.53592187  0.6136377  -1.65929074 -1.40584706]
 [ 2.70150301  1.74634788  1.46544034  2.1946403   0.41987974  3.12802546]
 ...
 [ 1.78316054  0.77307344  0.79831961  0.60110703  1.25154793  0.40770195]
 [ 1.78316054  0.77307344  0.79831961  0.60110703  1.25154793  0.40770195]
 [ 0.16417885  0.77307344  1.46544034  1.69096887  1.25154793 -1.40584706]]


## Data split into train and test & save

In [13]:
from sklearn.model_selection import train_test_split
import pickle

In [14]:
# Split and save X_desc_scaled
X_train, X_test, y_train, y_test = train_test_split(X_desc_scaled, y, test_size=0.2, random_state=42)
print(X_train.shape, y_train.shape)

# Save the data splits to a pickle file
with open("data_splits.pkl", "wb") as f:
    pickle.dump((X_train, X_test, y_train, y_test), f)

(1151, 6) (1151, 1)


## Additional Data: Morgan Fingerprints

In [15]:
from rdkit.Chem import AllChem

# Morgan fingerprints
def compute_fingerprint_np(mol, radius=2, nBits=512):
    return np.array(AllChem.GetMorganFingerprintAsBitVect(mol, radius, nBits))

fingerprint_list = [compute_fingerprint_np(mol) for mol in tqdm(mols)]
X_fps = np.vstack(fingerprint_list)         # shape: (n_samples, 512)

100%|██████████| 1439/1439 [00:01<00:00, 1351.60it/s]


In [16]:
# Check data
print(X_fps.shape)
print(X_fps)

(1439, 512)
[[0 1 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 1 0 ... 0 0 0]
 ...
 [1 1 0 ... 0 0 0]
 [1 1 0 ... 0 0 0]
 [1 0 0 ... 0 0 0]]


In [17]:
# Normalize X_fps
scaler_fps = StandardScaler()
X_fps_scaled = scaler_fps.fit_transform(X_fps)
print(X_fps_scaled)

[[-0.42139974  2.74131226 -0.19168598 ... -0.41221196 -0.19745675
  -0.17550582]
 [-0.42139974 -0.3647888  -0.19168598 ... -0.41221196 -0.19745675
  -0.17550582]
 [-0.42139974  2.74131226 -0.19168598 ... -0.41221196 -0.19745675
  -0.17550582]
 ...
 [ 2.3730437   2.74131226 -0.19168598 ... -0.41221196 -0.19745675
  -0.17550582]
 [ 2.3730437   2.74131226 -0.19168598 ... -0.41221196 -0.19745675
  -0.17550582]
 [ 2.3730437  -0.3647888  -0.19168598 ... -0.41221196 -0.19745675
  -0.17550582]]


In [18]:
# Split and save X_fps_scaled
X_train_fps, X_test_fps, y_train_fps, y_test_fps = train_test_split(X_fps_scaled, y, test_size=0.2, random_state=42)
print(X_train_fps.shape, y_train_fps.shape)

with open("data_splits_fps.pkl", "wb") as f:
    pickle.dump((X_train_fps, X_test_fps, y_train_fps, y_test_fps), f)

print("saved to pickle file")

(1151, 512) (1151, 1)
saved to pickle file
